go/worker/keymanager/status.go

```diff
@@ -0,0 +1,73 @@
+package keymanager
+
+import (
+	"context"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/worker/keymanager/api"
+)
+
+// GetStatus returns the key manager worker status.
+func (w *Worker) GetStatus(ctx context.Context) (*api.Status, error) {
+	var initialized, stopped bool
+	select {
+	case <-w.Initialized():
+		initialized = true
+	default:
+	}
+	select {
+	case <-w.Quit():
+		stopped = true
+	default:
+	}
+
+	var ss api.StatusState
+	switch {
+	case !w.enabled:
+		ss = api.StatusStateDisabled
+	case stopped:
+		ss = api.StatusStateStopped
+	case initialized:
+		ss = api.StatusStateReady
+	default:
+		ss = api.StatusStateStarting
+	}
+
+	var rid *common.Namespace
+	if w.runtime != nil {
+		id := w.runtime.ID()
+		rid = &id
+	}
+
+	rts := make([]common.Namespace, 0, len(w.clientRuntimes))
+	for rt := range w.clientRuntimes {
```

Comment:
Accessing this outside the `worker` goroutine will cause a potential data race as it is currently not protected by a lock. In this case it would probably be best to move the map constructor outside the worker and make sure to only access/manipulate it with the appropriate lock held.